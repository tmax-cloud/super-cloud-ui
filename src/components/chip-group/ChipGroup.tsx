import * as React from 'react';
import { Box, IconButton, ListItem, styled, SvgIcon, Typography } from '@mui/material';
import ThemeWrapper from '../../themes/ThemeWrapper';
import Tooltip from '../tooltip/Tooltip';

const StyledChipGroup = styled(Box)(({ theme }) => ({
  display: 'inline-flex',
  flaxWrap: 'wrap',
  alignItems: 'center',
  padding: `${theme.spaces.chipGroup.paddingTop} ${theme.spaces.chipGroup.paddingRight} ${theme.spaces.chipGroup.paddingLeft} ${theme.spaces.chipGroup.paddingBottom}`,
  backgroundColor: theme.palette.chipGroup.background,
  borderRadius: 3,
}));

const StyledChipGroupMain = styled(Box)(({ theme }) => ({
  display: 'flex',
  flex: 1,
  flexWrap: 'wrap',
  alignItems: 'baseline',
}));

const StyledChipGroupLabel = styled(Typography)(({ theme }) => ({
  maxWidth: '18ch',
  marginRight: theme.spaces.chipGroup.labelMarginRight,
  fontSize: theme.typography.chipGroup.fontSize,
}));

const StyledChipGroupList = styled(Box)(({ theme }) => ({
  display: 'inline-flex',
  flexWrap: 'wrap',
  alignItems: 'center',
  margin: 0,
  padding: 0,
  marginRight: theme.spaces.chipGroup.listMarginRight,
  marginBottom: theme.spaces.chipGroup.listMarginBottom,
}));

const StyledChipGroupListItem = styled(ListItem)(({ theme }) => ({
  display: 'inline-flex',
  width: 'auto',
  marginRight: theme.spaces.chipGroup.listItemMarginRight,
  marginBottom: theme.spaces.chipGroup.listItemMarginBottom,
}));

const StyledChipGroupClose = styled(Box)(({ theme }) => ({
  display: 'flex',
  alignItems: 'flex-start',
  marginTop: theme.spaces.chipGroup.closeMarginTop,
  marginBottom: theme.spaces.chipGroup.closeMarginBottom,
}));

const StyledChipGroupCloseButton = styled(IconButton)(({ theme }) => ({
  padding: `${theme.spaces.global.spacerXs} ${theme.spaces.global.spacerMd}`,
}));

const StyledChipGroupCloseIcon = styled(SvgIcon)(({ theme }) => ({
  color: theme.palette.chipGroup.icon,
  fontSize: theme.typography.chipGroup.fontSize,
  '&:hover': {
    color: theme.palette.chip.border,
  },
}));

const ChipGroupCloseButton = (props: ChipGroupCloseButtonProps) => {
  const { onDeleteAll } = props;
  return (
    <StyledChipGroupCloseButton disableRipple onClick={onDeleteAll}>
      <StyledChipGroupCloseIcon viewBox="0 0 352 512">
        <path d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z" transform=""></path>
      </StyledChipGroupCloseIcon>
    </StyledChipGroupCloseButton>
  );
};

const ChipGroup = (props: ChipGroupProps) => {
  const { categoryName, children, onDeleteAll } = props;
  const items = React.Children.toArray(children);
  const ref = React.useRef<HTMLDivElement>(null);
  const [showTooltip, setShowTooltip] = React.useState(false);

  React.useEffect(() => {
    if (ref.current) {
      const p = ref.current as HTMLParagraphElement;
      setShowTooltip(p.offsetWidth < p.scrollWidth);
    }
  }, [ref]);

  return (
    <>
      {items.length === 0 ? null : (
        <ThemeWrapper>
          <StyledChipGroup>
            <StyledChipGroupMain>
              {categoryName && (
                <Tooltip content={showTooltip && categoryName}>
                  <StyledChipGroupLabel ref={ref} noWrap>
                    {categoryName}
                  </StyledChipGroupLabel>
                </Tooltip>
              )}
              <StyledChipGroupList component="ul">
                {items?.map((item, index) => {
                  return (
                    <StyledChipGroupListItem key={`item-${index}`} disableGutters disablePadding>
                      {item}
                    </StyledChipGroupListItem>
                  );
                })}
              </StyledChipGroupList>
            </StyledChipGroupMain>
            {onDeleteAll && (
              <StyledChipGroupClose>
                <ChipGroupCloseButton onDeleteAll={onDeleteAll} />
              </StyledChipGroupClose>
            )}
          </StyledChipGroup>
        </ThemeWrapper>
      )}
    </>
  );
};

export interface ChipGroupCloseButtonProps {
  /**
   * Function that is called when clicking on the chip group close button.
   * If set, the delete icon will be shown.
   */
  onDeleteAll?: React.MouseEventHandler;
}
export interface ChipGroupBaseProps {
  /**
   * Category name text for the chip group category.
   */
  categoryName?: string;
  /**
   * Content rendered inside the chip group. Should be <Chip> elements.
   */
  children?: React.ReactNode;
}

export type ChipGroupProps = ChipGroupBaseProps & ChipGroupCloseButtonProps;

export default ChipGroup;
