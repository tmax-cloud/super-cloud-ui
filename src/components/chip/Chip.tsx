import * as React from 'react';
import { Chip as MuiChip, styled, SvgIcon, SxProps, Theme } from '@mui/material';
import { useTooltip } from '../../hooks/use-tooltip';
import ThemeWrapper from '../../themes/ThemeWrapper';
import Tooltip from '../tooltip/Tooltip';

const StyledChip = styled(MuiChip)(({ theme }) => ({
  '&.MuiChip-root': {
    borderColor: theme.palette.chip.border,
    borderRadius: 3,
    backgroundColor: theme.palette.chip.background,
    color: theme.palette.chip.text,
    fontSize: theme.typography.chip.fontSize,
  },
  '& .MuiChip-label': {
    padding: `${theme.spaces.chip.paddingTop} ${theme.spaces.chip.paddingRight}  ${theme.spaces.chip.paddingBottom}  ${theme.spaces.chip.paddingLeft}`,
    maxWidth: '7.5rem',
  },
  '.Chip-deleteIcon': {
    margin: `${theme.spaces.chip.buttonMarginTop} ${theme.spaces.chip.buttonMarginRight}  ${theme.spaces.chip.buttonMarginBottom}  ${theme.spaces.chip.buttonMarginLeft}`,
    color: theme.palette.chip.icon,
    fontSize: theme.typography.chip.fontSize,
    '&:hover': {
      color: theme.palette.chip.border,
    },
  },
}));

const ChipDeleteIcon = (
  <SvgIcon aria-label="delete" viewBox="0 0 512 512" className="Chip-deleteIcon">
    <path d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z" transform=""></path>
  </SvgIcon>
);

const Chip = (props: ChipProps) => {
  const { label, ...rest } = props;
  const { ref, showTooltip } = useTooltip<HTMLDivElement>('firstElementChild');

  return (
    <ThemeWrapper>
      <Tooltip content={showTooltip && label}>
        <StyledChip ref={ref} label={label} deleteIcon={ChipDeleteIcon} variant="outlined" {...rest} />
      </Tooltip>
    </ThemeWrapper>
  );
};

export interface ChipProps {
  /**
   * The content of the component.
   */
  label?: React.ReactNode;
  /**
   * Callback fired when the delete icon is clicked.
   * If set, the delete icon will be shown.
   */
  onDelete?: React.EventHandler<any>;
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx?: SxProps<Theme>;
}

export default Chip;
