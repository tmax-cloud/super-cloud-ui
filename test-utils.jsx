// import React from 'react';
// import { render } from '@testing-library/react';
// 추후에 redux 추가하게 되면 redux Provider로 감싸주는 custom render 메서드 추가 해야됨.

export * from '@testing-library/react';
