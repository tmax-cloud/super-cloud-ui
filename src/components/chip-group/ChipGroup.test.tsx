import * as React from 'react';
import { render } from '../../../test-utils';
import ChipGroup from './ChipGroup';
import Chip from '../chip/Chip';

describe('Chip', () => {
  test('renders correctly', () => {
    const { container } = render(
      <ChipGroup>
        <Chip label="Chip 1" />
      </ChipGroup>,
    );
    expect(container).toMatchSnapshot();
  });
  test('chip group with category', () => {
    const { container } = render(
      <ChipGroup categoryName="category">
        <Chip label="Chip 1" />
      </ChipGroup>,
    );
    expect(container).toMatchSnapshot();
  });
  test('chip group with deletable category', () => {
    const { container } = render(
      <ChipGroup categoryName="category" onDeleteAll={() => {}}>
        <Chip label="Chip 1" />
      </ChipGroup>,
    );
    expect(container).toMatchSnapshot();
  });
  test('chip group with category and tooltip', () => {
    const { container } = render(
      <ChipGroup categoryName="A very long category name">
        <Chip label="Chip 1" />
      </ChipGroup>,
    );
    expect(container).toMatchSnapshot();
  });
  test('chip group will not render if no children passed', () => {
    const { container } = render(<ChipGroup />);
    expect(container.innerHTML).toHaveLength(0);
  });
});
