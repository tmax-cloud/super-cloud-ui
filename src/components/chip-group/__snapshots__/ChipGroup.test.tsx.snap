// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip chip group with category 1`] = `
<div>
  <div
    class="MuiBox-root css-2ps56d"
    role="group"
  >
    <div
      class="MuiBox-root css-79yqfn"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-2zmal-MuiTypography-root"
      >
        category
      </p>
      <ul
        class="MuiBox-root css-14zi1m1"
      >
        <li
          class="MuiListItem-root css-1o8m0dh-MuiListItem-root"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-1qfz615-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Chip 1
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Chip chip group with category and tooltip 1`] = `
<div>
  <div
    class="MuiBox-root css-2ps56d"
    role="group"
  >
    <div
      class="MuiBox-root css-79yqfn"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-2zmal-MuiTypography-root"
      >
        A very long category name
      </p>
      <ul
        class="MuiBox-root css-14zi1m1"
      >
        <li
          class="MuiListItem-root css-1o8m0dh-MuiListItem-root"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-1qfz615-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Chip 1
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Chip chip group with deletable category 1`] = `
<div>
  <div
    class="MuiBox-root css-2ps56d"
    role="group"
  >
    <div
      class="MuiBox-root css-79yqfn"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-2zmal-MuiTypography-root"
      >
        category
      </p>
      <ul
        class="MuiBox-root css-14zi1m1"
      >
        <li
          class="MuiListItem-root css-1o8m0dh-MuiListItem-root"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-1qfz615-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Chip 1
            </span>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="MuiBox-root css-1b5ye64"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-cr3bmk-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1dblj6t-MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 352 512"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            transform=""
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Chip renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-2ps56d"
    role="group"
  >
    <div
      class="MuiBox-root css-79yqfn"
    >
      <ul
        class="MuiBox-root css-14zi1m1"
      >
        <li
          class="MuiListItem-root css-1o8m0dh-MuiListItem-root"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-1qfz615-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Chip 1
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
